QandA/06-functions.Rmd

```diff
@@ -123,53 +322,66 @@ h05(ls())
 ```
 :::
 
-:::TODO
-XXX
-:::
+The difference is where the promise is created. `ls()` is always evaluated inside `h05` when `x` is evaluated. The difference is the environment. When `ls()` is provided as an explicit parameter, `x` is a promise whose environment is the global environment. When `ls()` is a default parameter, it is evaluated in the local environment where it is used.
 
-## 6.5.3 Missing arguments {-}
+Hypothesis: does nesting `ls()` in `h05` first evaluate `ls()` then evaluate `h05()` ?
 
-:::question
-Comparing the default sample function:
-```{r}
-sample <- function (x, size, replace = FALSE, prob = NULL) 
-{
-    if (length(x) == 1L && is.numeric(x) && is.finite(x) && x >= 
-        1) {
-        if (missing(size))
-          # if you don't supply the sample size
-          # you will just get random samples
-          # the length of the supplied vector
-            size <- x
-        sample.int(x, size, replace, prob)
-    }
-    else {
-        if (missing(size)) 
-            size <- length(x)
-        x[sample.int(length(x), size, replace, prob)]
-    }
+```{r, eval=FALSE}
+library(magrittr)
+h05 <- function(x = {y <- 4;ls()}) {
+   a <- 1
+   x
 }
+```
 
-sample(1:10)
+```{r, eval=FALSE}
+h05()
 ```
 
-To using Hadley's recommended `NULL` version:
+```
+[1] "a" "x" "y"
+```
 
-```{r}
-my_sample_null <- function(x, size = NULL, replace = FALSE, prob = NULL) {
-  size <- size %||% length(x)
-  x[sample.int(length(x), size, replace = replace, prob = prob)]
-}
+```{r, eval=FALSE}
+ls()
+```
 
-my_sample_null(1:10)
+```
+[1] "h05"
 ```
 
-The second function is obviously more concise, but why is Hadley recommending we we steer clear of `missing()`? Is this just for code read-ability or is there something inherent in the behavior of `missing` we need to consider that we don't when using `NULL`?
-:::
+```{r, eval=FALSE}
+{y <- 4;ls()} %>% h05()
+```
 
-:::TODO
-XXX
-:::
+```
+[1] "h05" "y"  
+```
+
+```{r, eval=FALSE}
+ls()
+```
+
+```
+[1] "h05" "y" 
+```
+
+```{r, eval=FALSE}
+h05({x <- 5;ls()})
+```
+
+```
+[1] "h05" "x"   "y" 
+```
+
+```{r, eval=FALSE}
+ls()
+```
+```
+[1] "h05" "x"   "y" 
+```
+
+Notice in all of the latter calls, a is not returned - so it's not evaluating `ls()` inside of the function.
```

Comment:
Hmm, this was one of my answers that Kent edited - I think for understanding purposes, it amounts to the same thing?

Reply:
Ohh, I see Kent's comment a little earlier, okay that makes sense